google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -371,14 +401,47 @@ void processReceivedMessages(List<ReceivedMessage> messages) {
         // we want to eventually
         // totally expire so that pubsub service sends us the message again.
         continue;
+      } else {
+        outstandingBatch.add(outstandingMessage);
       }
-      outstandingBatch.add(new OutstandingMessage(message, ackHandler));
       pendingReceipts.add(ackRequestData);
     }
-
     processBatch(outstandingBatch);
   }
 
+  synchronized void notifyAckSuccess(AckRequestData ackRequestData) {
+
+    if (outstandingReceipts.containsKey(ackRequestData.getAckId())) {
+      ReceiptCompleteData receiptCompleteData = outstandingReceipts.get(ackRequestData.getAckId());
```

Comment:
We don't use receiptCompleteData or outstandingMessage, so let's stop storing these values locally, and just have a single line that sets the receipt to complete:

outstandingReceipts.get(ackRequestData.getAckId()).notifyReceiptComplete();